crates/astria-sequencer/src/transaction/query.rs

```diff
@@ -0,0 +1,106 @@
+use astria_core::{
+    generated::protocol::transaction::v1alpha1::UnsignedTransaction as RawUnsignedTransaction,
+    protocol::{
+        abci::AbciErrorCode,
+        transaction::v1alpha1::UnsignedTransaction,
+    },
+};
+use cnidarium::Storage;
+use prost::Message as _;
+use tendermint::abci::{
+    request,
+    response,
+};
+
+use crate::{
+    state_ext::StateReadExt,
+    transaction::checks::get_fees_for_transaction,
+};
+
+pub(crate) async fn transaction_fee_request(
+    storage: Storage,
+    request: request::Query,
+    _params: Vec<(String, String)>,
+) -> response::Query {
+    use astria_core::protocol::transaction::v1alpha1::TransactionFeeResponse;
+
+    let tx = match preprocess_request(&request) {
+        Ok(tx) => tx,
+        Err(err_rsp) => return err_rsp,
+    };
+
+    // use latest snapshot, as this is a query for a transaction fee
+    let snapshot = storage.latest_snapshot();
+    let height = match snapshot.get_block_height().await {
+        Ok(height) => height,
+        Err(err) => {
+            return response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
+                log: format!("failed getting block height: {err:#}"),
+                ..response::Query::default()
+            };
+        }
+    };
+
+    let fees = match get_fees_for_transaction(&tx, &snapshot).await {
+        Ok(fees) => fees,
+        Err(err) => {
+            return response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
+                log: format!("failed calculating fees for provided transaction: {err:?}"),
+                ..response::Query::default()
+            };
+        }
+    };
+
+    let fees = fees.into_iter().collect();
+
+    let resp = TransactionFeeResponse {
+        height,
+        fees,
+    };
+
+    let payload = resp.into_raw().encode_to_vec().into();
+
+    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
+    response::Query {
+        code: 0.into(),
+        key: request.path.clone().into_bytes().into(),
+        value: payload,
+        height,
+        ..response::Query::default()
+    }
+}
+
+fn preprocess_request(request: &request::Query) -> Result<UnsignedTransaction, response::Query> {
+    let tx = match RawUnsignedTransaction::decode(&*request.data) {
+        Ok(tx) => tx,
+        Err(err) => {
+            return Err(response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
```

Comment:
This doesn't seem like an internal server error - it points to the request being malformed.  I'm not sure if any of the existing `AbciErrorCode` variants really fit.  We need something like a `BAD_REQUEST` (equivalent to HTTP 400 code) variant I think.